src/java/apiview-java-processor/src/main/java/com/azure/tools/apiview/processor/diagnostics/rules/BuilderTraitsDiagnosticRule.java

```diff
@@ -0,0 +1,137 @@
+import static com.azure.tools.apiview.processor.model.DiagnosticKind.ERROR;
+import static com.azure.tools.apiview.processor.model.DiagnosticKind.WARNING;
+
+/**
+ * This diagnostic ensures that builder traits are applied correctly against client builders.
+ */
+public class BuilderTraitsDiagnosticRule implements DiagnosticRule {
+    private final Map<String, List<TraitMethod>> traits;
+
+    public BuilderTraitsDiagnosticRule() {
+        traits = new HashMap<>();
+        traits.put("AzureKeyCredentialTrait", Arrays.asList(
+            new TraitMethod("credential", "AzureKeyCredential")
+        ));
+        traits.put("AzureNamedKeyCredentialTrait", Arrays.asList(
+            new TraitMethod("credential", "AzureNamedKeyCredential")
+        ));
+        traits.put("AzureSasCredentialTrait", Arrays.asList(
+            new TraitMethod("credential", "AzureSasCredential")
+        ));
+        traits.put("TokenCredentialTrait", Arrays.asList(
+            new TraitMethod("credential", "TokenCredential")
+        ));
+        traits.put("ConfigurationTrait", Arrays.asList(
+            new TraitMethod("configuration", "Configuration")
+        ));
+        traits.put("ConnectionStringTrait", Arrays.asList(
+            new TraitMethod("connectionString", "String")
+        ));
+        traits.put("EndpointTrait", Arrays.asList(
+            new TraitMethod("endpoint", "String")
+        ));
+        traits.put("HttpTrait", Arrays.asList(
+            new TraitMethod("httpClient", "HttpClient"),
+            new TraitMethod("pipeline", "HttpPipeline"),
+            new TraitMethod("addPolicy", "HttpPipelinePolicy"),
+            new TraitMethod("retryOptions", "RetryOptions"),
+            new TraitMethod("httpLogOptions", "HttpLogOptions"),
+            new TraitMethod("clientOptions", "ClientOptions")
+        ));
+        traits.put("AmqpTrait", Arrays.asList(
+            new TraitMethod("retryOptions", "AmqpRetryOptions"),
+            new TraitMethod("transportType", "AmqpTransportType"),
+            new TraitMethod("proxyOptions", "ProxyOptions"),
+            new TraitMethod("clientOptions", "ClientOptions")
+        ));
+    }
+
+    @Override
+    public void scanIndividual(final CompilationUnit cu, final APIListing listing) {
+        // If the CU has a @ServiceClientBuilder annotation:
+        //   Look for methods with names of methods we know should belong to the trait:
+        //     Ensure the matching trait is stated as being implemented. If the method exists but the trait interface is
+        //     not defined as being implemented, add a warning diagnostic that the trait should be implemented.
+        //     If the method name exists, but it has different parameters that what we expect, add a warning diagnostic
+        //     suggesting to use the trait.
+
+        cu.getTypes().forEach(type -> {
+            if (!type.isAnnotationPresent("ServiceClientBuilder")) return;
+
+            // iterate through trait-by-trait
```

Comment:
Thoughts on inverting this by collecting all the methods on the class then iterating over the traits and determining if they should be added to `implements`.

Reply:
If I was concerned about perf, I would consider this, but the code is not on any perf critical path, so I'm at peace with how it is right now.